server/utils.go

```diff
@@ -174,3 +185,196 @@ func (w *withCode) Format(s fmt.State, verb rune) {
+	operationsMap := map[string][]map[string]map[string]oputil.DataAction{}
+
+	apiResourcesList, err := fetchApiResources()
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to fetch list of api-resources from apiserver.")
+	}
+
+	operationsList, err := fetchDataActionsList(environment, clusterType, loginURL, tenantId)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to fetch operations from Azure.")
+	}
+
+	for _, resList := range apiResourcesList {
+		group := "v1" // core api group
+		if resList.GroupVersion != "" && resList.GroupVersion != "v1" {
+			group = strings.Split(resList.GroupVersion, "/")[0]
+		}
+
+		if len(resList.APIResources) == 0 {
+			continue
+		}
+
+		for _, apiResource := range resList.APIResources {
+			if strings.Contains(apiResource.Name, "/") {
+				continue
+			}
+
+			actionId := clusterType
+			if group != "v1" {
+				actionId = path.Join(actionId, group)
+			}
+
+			resourceName := apiResource.Name
+
+			actionId = path.Join(actionId, resourceName)
+
+			for _, operation := range operationsList {
+				if strings.Contains(operation.Name, actionId) {
+					opNameArr := strings.Split(operation.Name, "/")
+
+					if group != "v1" {
+						// extra validation to make sure groups are the same
+						if !(group == opNameArr[2]) {
+							continue
+						}
+					} else {
+						// make sure resources are the same for core apigroup
+						if !(resourceName == opNameArr[2]) {
+							continue
+						}
+					}
+
+					verb := opNameArr[len(opNameArr)-1]
+					if verb == "action" {
+						verb = path.Join(opNameArr[len(opNameArr)-2], opNameArr[len(opNameArr)-1])
+					}
+
+					da := oputil.DataAction{
+						ActionInfo: oputil.AuthorizationActionInfo{
+							IsDataAction: true,
+						},
+						IsNamespacedResource: apiResource.Namespaced,
+					}
+					da.ActionInfo.AuthorizationEntity.Id = operation.Name
+
+					if operationsMap[group] == nil {
+						operationsMap[group] = []map[string]map[string]oputil.DataAction{}
+					}
+					resourceAndVerb := map[string]map[string]oputil.DataAction{}
+					resourceAndVerb[resourceName] = map[string]oputil.DataAction{}
+					resourceAndVerb[resourceName][verb] = da
+					operationsMap[group] = append(operationsMap[group], resourceAndVerb)
+				}
+			}
+		}
+	}
+
+	klog.V(5).Infof("Operations list: %v", operationsMap)
+
+	return operationsMap, nil
+}
+
+func fetchApiResources() ([]*metav1.APIResourceList, error) {
+	// creates the in-cluster config
+	klog.V(5).Infof("Fetch apiresources list")
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, errors.Wrap(err, "Error building kubeconfig")
+	}
+
+	kubeclientset, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error building kubernetes clientset")
+	}
+
+	apiresourcesList, err := kubeclientset.Discovery().ServerPreferredResources()
+	if err != nil {
+		return nil, err
+	}
+
+	klog.V(5).Infof("List of ApiResources fetched from apiserver: %v", apiresourcesList)
```

Comment:
there should be a String() or Marshal() that will look much better than %v

Reply:
done